packages/redux-dynamic-modules-observable/src/EpicManager.ts

```diff
@@ -1,81 +1,127 @@
+         * minor worry.
+         */
+        add(epics: Epic[] = []) {
+            epics.forEach(epic => {
+                const epicKey = epic.toString();
+                // Check if epic already exists
+                if (!runningEpics.hasOwnProperty(epicKey)) {
+                    const replaceableWrapper = createReplaceableWrapper();
+                    // we put replaceable Observable wrapper into epicMiddleware
+                    epicMiddleware.run(replaceableWrapper);
+                    // let's roll epic. Here we make epic run truly
+                    replaceableWrapper.replaceWith(epic);
+                    /**
+                     * We store the reference of replaceableWrapper, so we can check if it exists next time
+                     *
+                     * Is there a limit on length of the key (string) in JS object?
+                     * See https://stackoverflow.com/questions/13367391/is-there-a-limit-on-length-of-the-key-string-in-js-object
+                     */
+                    runningEpics[epicKey] = replaceableWrapper;
+                }
+                /**
+                 * We follow practice on official document https://redux-dynamic-modules.js.org/#/reference/ModuleCounting
+                 * So we use RefCounter to determine when we should remove epic
+                 */
+                epicRefCounter.add(epic);
             });
         },
-        remove: (epics: Epic[]) => {
-            if (!epics) {
-                return;
-            }
+        /**
+         * Remove epics
+         * Actually it will replace the real epic with a empty epic
+         *
+         * __Note:__
+         * Under some circumstances here https://redux-observable.js.org/docs/recipes/AddingNewEpicsAsynchronously.html
+         * We can't do a actual replacement.
+         * But we can try to replace real epic with empty epic, it works as we expected. This benefit is given by rxjs switchMap
+         */
+        remove(epics: Epic[] = []) {
+            epics.forEach(epic => {
+                epicRefCounter.remove(epic);
 
-            const epicNameMap = epics.reduce((p, e) => {
-                p[e.name] = e;
-                return p;
-            }, {});
-
-            epics.forEach(e => {
-                epicRefCounter.remove(e.name);
-            });
-
-            runningEpics = runningEpics.filter(e => {
-                !!epicNameMap[e.name] && epicRefCounter.getCount(e.name) !== 0;
+                const epicKey = epic.toString();
+                const replaceableWrapper = runningEpics[epicKey];
+                // Check if no module reference epic, we will remove epic
+                if (replaceableWrapper && !epicRefCounter.getCount(epic)) {
+                    // Replace the epic with empty epic, so no more unnecessary logic can cause any side effects.
+                    replaceableWrapper.replaceWith(emptyEpic);
```

Comment:
Does this mean that if I were to add/remove a module 100 times, there would be 100 empty epics? 
I am not very familiar with redux-observable, will that cause a performance problem?

Reply:
Actually it's all up to how the operator `switchMap` of RxJS works.  

To clarify

1. `replaceableWrapper` should be same as `runningEpics` in quantity.
2. One `replaceableWrapper` should maintain for one running Epic.
3. `switchMap` helps `replaceableWrapper` replace previous Epic with new Epic.

So If we replace 100 times, the last one will be kept.

The source text segment of Epic (it is a function) is used as key in `runningEpics`. It won't add one more instance in runtime environment unless the key changed by some way. If the source text segment of Epic is changed, it will be a new Epic and create a new `replaceableWrapper`.

`replaceableWrapper` behaves like `switchMap`. You may need read this quote from [learnrxjs.io](https://www.learnrxjs.io/operators/transformation/switchmap.html) for more information.
> ...a long lived inner observable could cause memory leaks, for instance if you used mergeMap with an interval and forgot to properly dispose of inner subscriptions. Remember, switchMap maintains only one inner subscription at a time, this can be seen clearly...

Reply:
I opened a new [PR](https://github.com/microsoft/redux-dynamic-modules/pull/115) to reuse replaceableWrapper